vortex-array/src/encoding.rs

```diff
@@ -1,24 +1,141 @@
 //! Traits and types to define shared unique encoding identifiers.
 
-use std::fmt::Debug;
+use std::any::Any;
+use std::fmt::{Debug, Display, Formatter};
+use std::sync::Arc;
 
 use arcref::ArcRef;
+use vortex_buffer::ByteBuffer;
+use vortex_dtype::DType;
+use vortex_error::VortexResult;
 
-use crate::vtable::{EncodingVTable, VTableRef};
-use crate::{Array, DeserializeMetadata, SerializeMetadata};
+use crate::serde::ArrayParts;
+use crate::vtable::{EncodeVTable, SerdeVTable, VTable};
+use crate::{Array, ArrayContext, ArrayRef, Canonical, DeserializeMetadata, IntoArray};
 
 /// EncodingId is a globally unique name of the array's encoding.
 pub type EncodingId = ArcRef<str>;
 
+pub type EncodingRef = ArcRef<dyn Encoding>;
+
 /// Marker trait for array encodings with their associated Array type.
-pub trait Encoding: 'static + Send + Sync + EncodingVTable + Debug {
-    type Array: Array;
-    type Metadata: SerializeMetadata + DeserializeMetadata + Debug;
+pub trait Encoding: 'static + private::Sealed + Send + Sync + Debug {
```

Comment:
this feels counter-intuitive to me, but I guess I get it.

Reply:
Another way to look at this is... I needed some encoding-specific data. So I renamed Encoding to VTable, and added a new Encoding trait!